mate-session-manager/doc/man/mate-session-save.1
.\"
.\" mate-session-save manual page.
.\" (C) 2000 Miguel de Icaza (miguel@helixcode.com)
.\" (C) 2009-2010 Vincent Untz (vuntz@gnome.org)
.\"
.TH MATE-SESSION-SAVE 1 "MATE"
.SH NAME
mate-session-save \- End or save the current MATE session
.SH SYNOPSIS
.B mate-session-save [\-\-logout] [\-\-force\-logout] [\-\-logout\-dialog] [\-\-shutdown\-dialog] [\-\-gui] [\-\-kill [\-\-silent]]
.SH DESCRIPTION
The \fImate-session-save\fP program can be used from a MATE session to
save a snapshot of the currently running applications. This session will
be later restored at your next MATE session.
.PP
The \fB\-\-gui\fP option will report errors in dialog boxes instead of
printing to stderr.
.PP
If called with the \fB\-\-logout\fP option, the current MATE session will be
ended, unless logging out has been inhibited by an application. The
\fB\-\-force\-logout\fP option can be used to end the session regardless of
the inhibition state.
.PP
When the \fB\-\-logout\-dialog\fP option is given, the standard dialog
displaying logout options is displayed. When \fB\-\-shutdown\-dialog\fP
option is given, the standard dialog displaying shutdown options is
displayed.
.PP
The \fB\-\-kill\fP and \fB\-\-silent\fP options are deprecated. The
\fB\-\-kill\fP option is equivalent to the \fB\-\-logout\-dialog\fP
option. If \fB\-\-silent\fP is used with \fB\-\-kill\fR, then it will
behave as if \fB\-\-logout\fP was used.
.PP
The session is not saved when \fImate-session-save\fP is called with any of
the options ending the session.
.SH BUGS
If you find bugs in the \fImate-session-save\fP program, please report
these on https://bugzilla.gnome.org.
.SH SEE ALSO
.BR mate-session(1)


libmate/NEWS
libmate 2.32.0

mate-sound APIs now use libcanberra internally. That means that
mate_sound_connection_get() and will always return error (-1), and
mate_sound_sample_load also always return -1 but still cache
the sample using the libcanberra API.
libmate continues to link to libesound and libaudiofile for ABI
compatibility; use --disable-esd to remove this dependency.
(#603198, #603349)


libmate 2.30.0

* Enable silent build using AM_SILENT_RULES (Christian Persch)
* Put SlowKeys and BounceKeys default delays back to 300 (Willie Walker)
* Deprecate mate-sound (Christian Persch)
* Deprecate mate_[un]setenv (Christian Persch)
* Update the description of the picture_options key (Matthias Clasen)

* Updated translations
  - Khaled Hosny (ar)
  - astur (ast)
  - Alexander Shopov (bg)
  - Sadia Afroz (bn)
  - Jordi Mallach (ca)
  - Reşat SABIQ (crh)
  - Andre Klapper (cs)
  - Iestyn Pryce (cy)
  - Ask Hjorth Larsen (da)
  - Mario Blättermann (de)
  - Bakaoukas Nikolaos (el)
  - Thomas Thurman (en@shaw)
  - Bruce Cowan (en_GB)
  - Jorge González (es)
  - Mattias Põldaru (et)
  - Iñaki Larrañaga Murgoitio (eu)
  - Tommi Vainikainen (fi)
  - Claude Paroz (fr)
  - Fran Diéguez (gl)
  - Gabor Kelemen (hu)
  - Francesco Marletta (it)
  - Rimas Kudelis (lt)
  - Badral Sanligiin (mn)
  - Kjartan Maraas (nb)
  - Nils-Christoph Fiedler (nds)
  - Wouter Bolsterlee (nl)
  - A S Alam (pa)
  - Piotr Drąg (pl)
  - Duarte Loreto (pt)
  - Henrique P Machado (pt_BR)
  - Adi Roiban (ro)
  - Yuri Kozlov (ru)
  - Marcel Telka (sk)
  - Matej Urbančič (sl)
  - Горан Ракић (sr)
  - Goran Rakić (sr@latin)
  - Daniel Nylander (sv)
  - Maxim Dziumanenko (uk)
  - Clytie Siddall (vi)
  - Chao-Hsiung Liao (zh_HK)
  - Chao-Hsiung Liao (zh_TW)


libmate 2.28.0

Aron Xu (1):
      Corrected 5 strings.
Denis Arnaud (1):
      Updated breton translation and added br in LINGUAS
Lucian Adrian Grijincu (1):
      Updated Romanian translation
Miloš Popović (1):
      Updated Serbian translation
Rodrigo L. M. Flores (1):
      Updated Brazilian Portuguese mailing list address.

 
libmate 2.27.5

* Update default properties for menus_have_icons, buttons_have_icons and
  toolbar_style
* Port mate-sound.c to g_poll to fix portability for old OSX versions without
  poll.
* Update the address of the FSF by replacing old bits with new ones

* Updated translations
  - Jordi Mallach (ca)
  - Jordi Mallach (ca@valencia)
  - Khaled Hosny (ar)
  - Kjartan Maraas (nb)
  - Marcel Telka (sk)
  - Ray Wang (zh_CN)
  - Shankar Prasad (kn)
  - Sweta Kothari (gu)


libmate 2.26.0

* Updated translations
  (as, bn_IN, crh, cs, da, el, gl, hi, it, ja, lt, ml, mr,
   ro, ru, ta, te)

libmate 2.24.1

* Fix warnings about missing format specifiers (Kjartan)
* Remove #if 0'ed out code (Kjartan)
* Updated translations
  (ar, bg, en_GB, eu, fr, gu, lv, nl, or, pl, pt, th, tr,
   vi, zh_HK, zh_TW)

libmate 2.24.0.1

* Make thumbnail housekeeping less agressive by default
  (Michael Chudobiak)

libmate 2.24.0

* Updated translations
  (ar, da, gu, it, ro, ru, sr, ta, te)

libmate 2.23.92

* Port mate-open to GIO. (Jean-Yves Lefort)
* Build fix (Tor Lillqvist)

libmate 2.23.5

* Fixes for sound related issues (Jeff Steadfast)

libmate 2.23.4

* Updated translations.

libmate 2.23.3

* Code cleanups (Kjartan, Christian)
* Added schema to support the thumbnail cache cleaner (Michael J. Chudobiak)
* Plug leak (Stefan Kost)
* Add disable_application_handlers lockdown schema (William Jon McCann)
 
libmate 2.22.0

* Release 2.22.0
 
libmate 2.21.91

* Add schemas for tasks and calendar
* Use the GIO backend for the filechooser

libmate 2.21.90

* Set sound as enabled by default (Christian)
* Bypass a11y on win32 (Tor)
* Remove unused popt stuff (dmacks at netspace org) 
* Fix the toolbar icon size schema type (#401030)
* Win32 fixes
* SElinux fixes
* Warning fixes
* Translation updates

libmate 2.20.0

* Build fixes (Christian Persch)
* Set new background (Lucas Rocha, Vincent Untz)
* Fix module loading (Mariano Suárez-Alvarez, Christian Persch)
* Only depend on esd when esd is compiled in (Daniel Gryniewicz)
* Updated translations:
  - Albanian (Laurent Dhima)
  - Arabic (Khaled Hosny)
  - Assamese (Amitakhya Phukan)
  - Basque (Inaki Larranaga Murgoitio)
  - Bengali (Khandakar Mujahidul Islam)
  - Brazilian Portugese (Og Maciel)
  - Bulgarian (Alexander Shopov)
  - Canadian English (Adam Weinberger)
  - Catalan (Gil Forcada)
  - Danish (Kenneth Nielsen)
  - Estonian (Ivar Smolin)
  - Finnish (Ilkka Tuohela)
  - French (Stephane Raimbault)
  - Georgian (Vladimer Sichinava)
  - German (Hendrik Richter)
  - Gujarati (Ankit Patel)
  - Hungarian (Gabor Kelemen)
  - Italian (Francesco Marletta)
  - Japanese (Takeshi AIHANA)
  - Lithuanian (Žygimantas Beručka)
  - Macedonian (Jovan Naumovski)
  - Malayalam (Ani Peter)
  - Norwegian bokmål (Kjartan Maraas)
  - Polish (MATE Pl Team)
  - Portugese (Duarte Loreto)
  - Romaninan (Mişu Moldovan)
  - Russian (Nikolay V. Shmyrev)
  - Serbian (Goran Rakić)
  - Sinhala (Danishka Navin)
  - Spanish (Jorge Gonzalez)
  - Swedish (Daniel Nylander)
  - Thai (Theppitak Karoonboonyanan)
  - Ukrainian (Maxim Dziumanenko)
  - Vietnamese (Clytie Siddall)

libmate 2.19.1

* Don't initialise ALSA when sound is disabled (449249; Christian Persch)
* Translation updates: Runa Bhattacharjee (bn_IN), Hendrik Richter (de),
  Iñaki Larrañaga Murgoitio (eu), Ilkka Tuohela (fi), Ankit Patel (gu),
  Takeshi AIHANA (ja), Leonardo Fereira Fontenelle (pt_BR), Clytie Siddall (vi)

libmate 2.19.0

What's new since 2.18.0

* Rebuild the type builtins when the Makefile changes (Christian Persch)
* Make bug-buddy work with applets again (Matthias Clasen, #424949)
* Fix problems when building static libraries (Kjartan Maraas, #405352)
* Fix typos (Kenneth Nielsen, Kjartan Maraas, #429751)
* Add more information to docs for mate_program_init () (Christian Persch,
  #82943)
* Updated translations
  - mg: Thierry Randrianiriana
  - nb: Kjartan Maraas

mate-session-manager/tools/mate-session-save.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 * save-session.c - Small program to talk to session manager.

   Copyright (C) 1998 Tom Tromey
   Copyright (C) 2008 Red Hat, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.
*/

#include <config.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <glib/gi18n.h>
#include <gtk/gtk.h>

#include <dbus/dbus-glib.h>
#include <dbus/dbus-glib-lowlevel.h>

#define GSM_SERVICE_DBUS "org.mate.SessionManager"
#define GSM_PATH_DBUS "/org/mate/SessionManager"
#define GSM_INTERFACE_DBUS "org.mate.SessionManager"

enum {
	GSM_LOGOUT_MODE_NORMAL = 0,
	GSM_LOGOUT_MODE_NO_CONFIRMATION,
	GSM_LOGOUT_MODE_FORCE
};

/* True if killing. This is deprecated, but we keep it for compatibility
 * reasons. */
static gboolean kill_session = FALSE;

/* The real options that should be used now. They are not ambiguous. */
static gboolean logout = FALSE;
static gboolean force_logout = FALSE;
static gboolean logout_dialog = FALSE;
static gboolean shutdown_dialog = FALSE;

/* True if we should use dialog boxes */
static gboolean show_error_dialogs = FALSE;

/* True if we should do the requested action without confirmation */
static gboolean no_interaction = FALSE;

static char* session_name = NULL;

static GOptionEntry options[] = {
	{"logout", '\0', 0, G_OPTION_ARG_NONE, &logout, N_("Log out"), NULL},
	{"force-logout", '\0', 0, G_OPTION_ARG_NONE, &force_logout, N_("Log out, ignoring any existing inhibitors"), NULL},
	{"logout-dialog", '\0', 0, G_OPTION_ARG_NONE, &logout_dialog, N_("Show logout dialog"), NULL},
	{"shutdown-dialog", '\0', 0, G_OPTION_ARG_NONE, &shutdown_dialog, N_("Show shutdown dialog"), NULL},
	{"gui",  '\0', 0, G_OPTION_ARG_NONE, &show_error_dialogs, N_("Use dialog boxes for errors"), NULL},
	/* deprecated options */
	{"session-name", 's', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_STRING, &session_name, N_("Set the current session name"), N_("NAME")},
	{"kill", '\0', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &kill_session, N_("Kill session"), NULL},
	{"silent", '\0', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &no_interaction, N_("Do not require confirmation"), NULL},
	{NULL}
};

static void display_error(const char* message)
{
	if (show_error_dialogs && !no_interaction)
	{
		GtkWidget* dialog = gtk_message_dialog_new(NULL, 0, GTK_MESSAGE_ERROR, GTK_BUTTONS_CLOSE, "%s", message);

		/*gtk_window_set_default_icon_name (GTK_STOCK_SAVE);*/

		gtk_dialog_run(GTK_DIALOG(dialog));
		gtk_widget_destroy(dialog);
	}
	else
	{
		g_printerr("%s\n", message);
	}
}

static DBusGConnection* get_session_bus(void)
{
	GError* error = NULL;
	DBusGConnection* bus = dbus_g_bus_get(DBUS_BUS_SESSION, &error);

	if (bus == NULL)
	{
		g_warning("Couldn't connect to session bus: %s", error->message);
		g_error_free(error);
	}

	return bus;
}

static DBusGProxy* get_sm_proxy(void)
{
	DBusGConnection* connection;
	DBusGProxy* sm_proxy;

	connection = get_session_bus();

	if (connection == NULL)
	{
		display_error(_("Could not connect to the session manager"));
		return NULL;
	}

	sm_proxy = dbus_g_proxy_new_for_name(connection, GSM_SERVICE_DBUS, GSM_PATH_DBUS, GSM_INTERFACE_DBUS);

	if (sm_proxy == NULL)
	{
		display_error(_("Could not connect to the session manager"));
		return NULL;
	}

	return sm_proxy;
}

#if 0
static void set_session_name(const char* session_name)
{
	DBusGProxy* sm_proxy;
	GError* error;
	gboolean res;

	sm_proxy = get_sm_proxy();

	if (sm_proxy == NULL)
	{
		return;
	}

	error = NULL;
	res = dbus_g_proxy_call(sm_proxy, "SetName", &error, G_TYPE_STRING, session_name, G_TYPE_INVALID, G_TYPE_INVALID);

	if (!res)
	{
		if (error != NULL)
		{
			g_warning("Failed to set session name '%s': %s", session_name, error->message);
			g_error_free(error);
		}
		else
		{
			g_warning("Failed to set session name '%s'", session_name);
		}
	}

	if (sm_proxy != NULL)
	{
		g_object_unref(sm_proxy);
	}
}
#endif

static void do_logout(unsigned int mode)
{
	DBusGProxy* sm_proxy;
	GError* error;
	gboolean res;

	sm_proxy = get_sm_proxy();

	if (sm_proxy == NULL)
	{
		return;
	}

	error = NULL;
	res = dbus_g_proxy_call(sm_proxy, "Logout", &error, G_TYPE_UINT, mode, G_TYPE_INVALID, G_TYPE_INVALID);

	if (!res)
	{
		if (error != NULL)
		{
			g_warning("Failed to call logout: %s", error->message);
			g_error_free(error);
		}
		else
		{
			g_warning("Failed to call logout");
		}
	}

	if (sm_proxy != NULL)
	{
		g_object_unref(sm_proxy);
	}
}

static void do_shutdown_dialog(void)
{
	DBusGProxy* sm_proxy;
	GError* error;
	gboolean res;

	sm_proxy = get_sm_proxy();

	if (sm_proxy == NULL)
	{
		return;
	}

	error = NULL;
	res = dbus_g_proxy_call(sm_proxy, "Shutdown", &error, G_TYPE_INVALID, G_TYPE_INVALID);

	if (!res)
	{
		if (error != NULL)
		{
			g_warning("Failed to call shutdown: %s", error->message);
			g_error_free(error);
		}
		else
		{
			g_warning("Failed to call shutdown");
		}
	}

	if (sm_proxy != NULL)
	{
		g_object_unref(sm_proxy);
	}
}

int main(int argc, char* argv[])
{
	GError* error;
	int conflicting_options;

	/* Initialize the i18n stuff */
	bindtextdomain(GETTEXT_PACKAGE, LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);

	error = NULL;

	if (!gtk_init_with_args(&argc, &argv, NULL, options, NULL, &error))
	{
		g_warning("Unable to start: %s", error->message);
		g_error_free(error);
		exit(1);
	}

	conflicting_options = 0;

	if (kill_session)
	{
		conflicting_options++;
	}

	if (logout)
	{
		conflicting_options++;
	}

	if (force_logout)
	{
		conflicting_options++;
	}

	if (logout_dialog)
	{
		conflicting_options++;
	}

	if (shutdown_dialog)
	{
		conflicting_options++;
	}

	if (conflicting_options > 1)
	{
		display_error(_("Program called with conflicting options"));
	}

	if (kill_session)
	{
		if (no_interaction)
		{
			force_logout = TRUE;
		}
		else
		{
			logout_dialog = TRUE;
		}
	}

	if (logout)
	{
		do_logout(GSM_LOGOUT_MODE_NO_CONFIRMATION);
	}
	else if (force_logout)
	{
		do_logout(GSM_LOGOUT_MODE_FORCE);
	}
	else if (logout_dialog)
	{
		do_logout(GSM_LOGOUT_MODE_NORMAL);
	}
	else if (shutdown_dialog)
	{
		do_shutdown_dialog();
	}

	return 0;
}
